src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ImportService.cs

```diff
@@ -1,163 +1,126 @@
 
-            await _tableStorageService.CreateOrUpdateEntity(DatafileImportsTableName,
-                new DatafileImport(releaseId.ToString(), dataFileName, 0, importReleaseMessage,
-                    IStatus.FAILED, JsonConvert.SerializeObject(errors)));
-        }
+            if (import == null)
+            {
+                return new ImportViewModel
+                {
+                    Status = ImportStatus.NOT_FOUND
+                };
+            }
 
-        private async Task UpdateImportTableRow(Guid releaseId, string dataFileName, int numberOfRows, ImportMessage message)
-        {
-            await _tableStorageService.CreateOrUpdateEntity(DatafileImportsTableName,
-                new DatafileImport(releaseId.ToString(), dataFileName, numberOfRows,
-                    JsonConvert.SerializeObject(message), IStatus.QUEUED));
+            return new ImportViewModel
+            {
+                Errors = import.Errors.Select(error => error.Message).ToList(),
+                PercentageComplete = import.PercentageComplete(),
+                StagePercentageComplete = import.StagePercentageComplete,
+                NumberOfRows = import.Rows,
+                Status = import.Status
+            };
         }
 
-        private ImportMessage BuildMessage(Guid subjectId, 
-            string dataFileName,
-            string metaFileName,
-            Guid releaseId,
-            string zipFileName = "")
+        public async Task Import(Guid subjectId, File dataFile, File metaFile, IFormFile formFile)
         {
-            var release = _context.Releases
-                .Where(r => r.Id.Equals(releaseId))
-                .Include(r => r.Publication)
-                .ThenInclude(p => p.Topic)
-                .ThenInclude(t => t.Theme)
-                .FirstOrDefault();
-
-            var importMessageRelease = _mapper.Map<Release>(release);
-
-            return new ImportMessage
+            var import = await _importRepository.Add(new Import
             {
+                Created = DateTime.UtcNow,
+                FileId = dataFile.Id,
+                MetaFileId = metaFile.Id,
                 SubjectId = subjectId,
-                DataFileName = dataFileName,
-                MetaFileName = metaFileName,
-                Release = importMessageRelease,
-                NumBatches = 1,
-                BatchNo = 1,
-                ArchiveFileName = zipFileName
-            };
-        }
-    }
+                Rows = CalculateNumberOfRows(formFile.OpenReadStream()),
+                Status = ImportStatus.QUEUED
+            });
 
-    public class ValidationError
-    {
-        public string Message { get; set; }
+            await _queueService.AddMessageAsync(ImportsPendingQueue, new ImportMessage(import.Id));
+        }
 
-        public ValidationError(string message)
+        public async Task ImportZip(Guid subjectId, File dataFile, File metaFile, File zipFile)
```

Comment:
They could be but the current test setup of the upload methods in `ReleaseDataFileService` verifiy the data file is passed to the data file argument, the meta file is passed to the meta file argument (and for the zip upload the zip file is passed to the zip file argument). It does this by depending on fields of the File, i.e. type and filename.

As it stands because we don't know the file id created by the file repository, these checks would then become It.IsAny<Guid>() which would pass if the params were in the wrong order.

I could change this and make the test class mock the FileRepository so we can control these Id's but I prefer to use the real repository class where possible, which in this case it is because it only depends on the database and we are setting up the database context.

I think the best thing to do in future would be too look at setting up the polymorphic types with table-per-hierarchy (TPH) or table-per-type (TPT) configuration using the Type as the discriminator. This would give us classes like `DataFile` `MetaFile` `ZipFile` so there is no possible way that they can be mixed up, and it would also be obvious then that fields like `Replacing` and `Source` only apply to `DataFile`.
https://docs.microsoft.com/en-us/ef/core/modeling/inheritance